stage0/src/alloc.rs

```diff
@@ -0,0 +1,125 @@
+//
+// Copyright 2023 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use core::{
+    alloc::Layout,
+    mem::MaybeUninit,
+    ptr::{write, NonNull},
+    sync::atomic::{AtomicUsize, Ordering},
+};
+use spinning_top::Spinlock;
+
+struct Inner<const N: usize> {
+    index: AtomicUsize,
+    storage: MaybeUninit<[u8; N]>,
+}
+
+impl<const N: usize> Inner<N> {
+    const fn new() -> Self {
+        Self {
+            index: AtomicUsize::new(0),
+            storage: MaybeUninit::uninit(),
+        }
+    }
+}
+
+/// Basic bump allocator that never deallocates memory.
+pub struct Allocator<const N: usize> {
+    inner: Spinlock<Inner<N>>,
+}
+
+impl<const N: usize> Allocator<N> {
+    pub const fn uninit() -> Self {
+        Self {
+            inner: Spinlock::new(Inner::new()),
+        }
+    }
+
+    pub fn allocate(&self, layout: Layout) -> Option<NonNull<u8>> {
+        let mut inner = self.inner.lock();
+        let offset = inner.index.load(Ordering::Acquire);
+        let storage_ptr = inner.storage.as_mut_ptr() as *mut u8;
+
+        // We may need to allocate extra bytes to ensure proper alignment in memory.
+        let align = if (storage_ptr as usize + offset) % layout.align() > 0 {
+            layout.align() - ((storage_ptr as usize + offset) % layout.align())
+        } else {
+            0
+        };
+
+        if offset + align + layout.size() > N {
+            // memory exhausted
+            return None;
+        }
+
+        inner
+            .index
+            .fetch_add(align + layout.size(), Ordering::SeqCst);
+
+        // Safety: we've reserved memory from [offset, offset + align + size) so this will not
+        // exceed the bounds of `self.storage` and is not aliased.
+        NonNull::new(unsafe { storage_ptr.add(offset + align) })
+    }
+
+    pub fn leak<T>(&self, val: T) -> Option<&mut T> {
+        let ptr = self.allocate(Layout::for_value(&val))?.cast().as_ptr();
+
+        // Safety: we've successfully allocated enough memory that is of the correct size, therefore
+        // writing `val` to it is safe. We also ensure that the pointer is properly initialized and
+        // aligned, so dereferencing it as a reference is fine.
+        unsafe {
+            write(ptr, val);
+            ptr.as_mut()
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
```

Comment:
Maybe add a test where we 
- immediately exceed the memory bounds to document that we did init everything correctly, and
- alloc which fails, followed by another alloc which should succeed---if that is possible?

Reply:
Both of these can be added to one test :)